Ros 2 timestamp conversion issue with Euroc Dataset

Hello,

I was able to create a simple launch file and run MSCEqF however, there seems to be a timestamp conversion issue.

My launch is as follows:

```
from launch import LaunchDescription

import launch.actions
import launch_ros.actions

bag_path = '/euroc/bags/MH_01_easy'
bag_start_offset = '50'

def generate_launch_description():
    return LaunchDescription([
        launch_ros.actions.Node(
            package='msceqf',
            executable='msceqf_ros2',
            name='msceqf_ros2',
            output='screen',
            parameters=[
                {'config_filepath': '/workspaces/examples/euroc/config/config.yaml'},
                {'imu_topic': '/imu0'},
                {'cam_topic': '/cam0/image_raw'},
                ],            
        ),
        launch.actions.ExecuteProcess(
            cmd=['ros2', 'bag', 'play',
                f'{bag_path}',
                '--start-offset',
                f'{bag_start_offset}', 
                ],
            output='log'
        )
])
```

When I run it as is I get:

```
[msceqf_ros2-1] -9.99881e-05  1.49672e-06  3.75619e-11            0            0            0            0            0            0 -1.06062e-07 -7.08519e-06 -9.10785e-06  8.88873e-07   5.9381e-05 -6.44214e-07   9.6804e-09  0.000999999 -3.75454e-07  8.13223e-07  2.44102e-08   6.5224e-06
[msceqf_ros2-1]   4.1403e-07  2.57155e-06  9.99661e-09            0            0            0            0            0            0 -1.83118e-07  3.02595e-08 -1.99835e-07  1.52692e-06 -2.60289e-07  3.71721e-06  2.57631e-06 -3.75454e-07  0.000900078 -4.63584e-09 -2.58279e-08  2.88579e-08
[msceqf_ros2-1] -8.13937e-07 -4.11921e-08 -2.06691e-10            0            0            0  1.48655e-10  9.99557e-09 -2.57744e-10  2.93743e-09  -5.7695e-08 -6.92223e-08 -2.44574e-08   4.8368e-07 -8.21589e-08 -5.32681e-08  8.13223e-07 -4.63584e-09  0.000900017  7.31417e-10  5.19378e-08
[msceqf_ros2-1] -3.89778e-08 -9.72828e-07 -6.49932e-10            0            0            0 -9.99881e-09  1.49672e-10  3.75619e-11  6.89932e-08 -2.82188e-09  8.61527e-08 -5.77726e-07  2.40847e-08 -1.40294e-06  -9.7296e-07  2.44102e-08 -2.58279e-08  7.31417e-10   0.00090002 -1.85544e-08
[msceqf_ros2-1] -6.49073e-06  2.16734e-06 -2.88706e-12            0            0            0   4.1403e-11  2.57155e-10  9.99661e-09 -1.53578e-07 -4.59936e-07 -7.82011e-07  1.28714e-06  3.85473e-06  2.94132e-06  2.06989e-06   6.5224e-06  2.88579e-08  5.19378e-08 -1.85544e-08  0.000900478.
[msceqf_ros2-1] terminate called after throwing an instance of 'std::runtime_error'
[msceqf_ros2-1]   what():  Time is out of dual 32-bit range
[ERROR] [msceqf_ros2-1]: process has died [pid 93588, exit code -6, cmd '/workspaces/install/msceqf/lib/msceqf/msceqf_ros2 --ros-args -r __node:=msceqf_ros2 --params-file /tmp/launch_params_5ln4wpic --params-file /tmp/launch_params_6d8eq35u --params-file /tmp/launch_params_9kjeom8r'].
```

So I made the following changes:

```
  static rclcpp::Time fromSec(double t)
  {
    // int64_t sec64 = static_cast<int64_t>(floor(t));
    // if (sec64 < 0 || sec64 > std::numeric_limits<uint32_t>::max())
    //   throw std::runtime_error("Time is out of dual 32-bit range");
    // uint32_t sec = static_cast<uint32_t>(sec64);
    // uint32_t nsec = static_cast<uint32_t>(std::round((t - sec) * 1e9));
    // sec += (nsec / 1000000000ul);
    // nsec %= 1000000000ul;
    // return rclcpp::Time(sec, nsec);

    const auto sec = std::chrono::duration<double>(t);
    const auto nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(sec);
    return rclcpp::Time(nsec.count());
  }

```

But then I get this behaviour
```


2.57631e-06 -3.75454e-07  0.000900078 -4.63584e-09 -2.58279e-08  2.88579e-08
[msceqf_ros2-1] -8.13937e-07 -4.11921e-08 -2.06691e-10            0            0            0  1.48655e-10  9.99557e-09 -2.57744e-10  2.93743e-09  -5.7695e-08 -6.92223e-08 -2.44574e-08   4.8368e-07 -8.21589e-08 -5.32681e-08  8.13223e-07 -4.63584e-09  0.000900017  7.31417e-10  5.19378e-08
[msceqf_ros2-1] -3.89778e-08 -9.72828e-07 -6.49932e-10            0            0            0 -9.99881e-09  1.49672e-10  3.75619e-11  6.89932e-08 -2.82188e-09  8.61527e-08 -5.77726e-07  2.40847e-08 -1.40294e-06  -9.7296e-07  2.44102e-08 -2.58279e-08  7.31417e-10   0.00090002 -1.85544e-08
[msceqf_ros2-1] -6.49073e-06  2.16734e-06 -2.88706e-12            0            0            0   4.1403e-11  2.57155e-10  9.99661e-09 -1.53578e-07 -4.59936e-07 -7.82011e-07  1.28714e-06  3.85473e-06  2.94132e-06  2.06989e-06   6.5224e-06  2.88579e-08  5.19378e-08 -1.85544e-08  0.000900478.
[msceqf_ros2-1] [ ERROR]: No IMU readings in between 913555501403636608.000000 and 963555501403636608.000000.
[msceqf_ros2-1] [ ERROR]: Unable to propagate. Empty propagation buffer.
[msceqf_ros2-1] [ ERROR]: Propagation failure.
[msceqf_ros2-1] [ WARNING]: Received IMU measurement older than actual state estimate. Discarding measurement.
[msceqf_ros2-1] [ WARNING]: Received IMU measurement older than actual state estimate. Discarding measurement.
[msceqf_ros2-1] [ WARNING]: Received IMU measurement older than actual state estimate. Discarding measurement.
[msceqf_ros2-1] [ WARNING]: Received IMU measurement older than actual state estimate. Discarding measurement.
```

Do the timestamps somehow get realigned in the code?

--- a/README.md
+++ b/README.md
@@ -139,7 +139,7 @@ extrinsics_std: [1.0e-2, 1.0e-2, 1.0e-2, 1.0e-2, 1.0e-2, 1.0e-2]
 intrinsics_std: [1.0, 1.0, 1.0, 1.0]
 
 # IMU noise statistics
-accelerometer_noise_density: 1.0-2
+accelerometer_noise_density: 1.0e-2
 accelerometer_random_walk:   1.0e-3
 gyroscope_noise_density: 1.0e-3
 gyroscope_random_walk:   1.0e-4

--- a/wrappers/ros/ros2/include/msceqf_ros.hpp
+++ b/wrappers/ros/ros2/include/msceqf_ros.hpp
@@ -88,14 +88,9 @@ private:
    */
   static rclcpp::Time fromSec(double t)
   {
-    int64_t sec64 = static_cast<int64_t>(floor(t));
-    if (sec64 < 0 || sec64 > std::numeric_limits<uint32_t>::max())
-      throw std::runtime_error("Time is out of dual 32-bit range");
-    uint32_t sec = static_cast<uint32_t>(sec64);
-    uint32_t nsec = static_cast<uint32_t>(std::round((t - sec) * 1e9));
-    sec += (nsec / 1000000000ul);
-    nsec %= 1000000000ul;
-    return rclcpp::Time(sec, nsec);
+    const auto sec = std::chrono::duration<double>(t);
+    const auto nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(sec);
+    return rclcpp::Time(nsec.count());
   }
 
   std::shared_ptr<rclcpp::Node> node_; //<! ROS node

--- a/wrappers/ros/ros2/source/msceqf_ros.cpp
+++ b/wrappers/ros/ros2/source/msceqf_ros.cpp
@@ -74,7 +74,7 @@ void MSCEqFRos::callback_image(const sensor_msgs::msg::Image::SharedPtr msg)
 
   msceqf::Camera cam;
 
-  cam.timestamp_ = cv_ptr->header.stamp.sec + 1.0e9 * cv_ptr->header.stamp.nanosec;
+  cam.timestamp_ = cv_ptr->header.stamp.sec + 1.0e-9 * cv_ptr->header.stamp.nanosec;
   cam.image_ = cv_ptr->image.clone();
 
   {
@@ -88,7 +88,7 @@ void MSCEqFRos::callback_imu(const sensor_msgs::msg::Imu::SharedPtr msg)
 {
   msceqf::Imu imu;
 
-  auto timestamp = msg->header.stamp.sec + 1.0e9 * msg->header.stamp.nanosec;
+  auto timestamp = msg->header.stamp.sec + 1.0e-9 * msg->header.stamp.nanosec;
 
   imu.timestamp_ = timestamp;
   imu.ang_ << msg->angular_velocity.x, msg->angular_velocity.y, msg->angular_velocity.z;

